library/Network/Wai/Middleware/OpenTelemetry.hs

```diff
@@ -27,18 +23,13 @@ newOpenTelemetryWaiMiddleware = do
 --
 -- This is added automatically by our 'newOpenTelemetryWaiMiddleware'.
 openTelemetryMiddleware :: Middleware
-openTelemetryMiddleware app request respond = do
-  mTraceId <- getCurrentTraceId
-  case mTraceId of
-    Nothing -> app request respond
-    Just traceId -> do
-      let
-        traceIdInt = convertOpenTelemetryTraceIdToDatadogTraceId traceId
-        traceIdIntBS = BS8.pack $ show traceIdInt
+openTelemetryMiddleware app request respond =
+  withTraceIdContext $ addTraceIdHeader app request respond
 
-      withThreadContext ["trace_id" .= traceIdInt] $
-        addHeaders
-          [("X-Datadog-Trace-Id", traceIdIntBS)]
-          app
-          request
-          respond
+addTraceIdHeader :: Middleware
+addTraceIdHeader app request respond = do
+  mTraceId <- getCurrentTraceIdAsDatadog
```

Comment:
Did you consider a `withTraceIdContext :: (MonadIO m, MonadMask m) => (Maybe Word64 -> m a) -> m a` to avoid the awkward `getCurrentTraceIdAsDatadog` inside of `withTraceIdContext`?

Reply:
I did. It seemed to make other, more general use-cases more awkward to make the middleware use-case less awkward, which I thought was a poor trade-off.

Reply:
Perhaps something like...

```hs
withTraceIdContext :: m a -> m a
withTraceIdContext f = withTraceIdContext' $ \_ -> f

withTraceIdContext' :: (Maybe Word64 -> m a) -> m a
withTraceIdContext'
```

I just wonder what value this brings to make for the fact now we have to talk about naming this stuff.

Reply:
Since I can extract that later without braking compatibility, I'm going to merge as is, but we can keep talking.

Reply:
Meh, I think it's going to be ugly either way. I'd probably feel less confused seeing `withTraceIdContext $ \_ -> do ...` then the `withTraceIdContext $ do { mTraceId  <- withTraceIdContext; ...}` but I don't feel strongly at all.

Reply:
That's the rub though. `withTraceIdContext $ \_` implies `(Maybe TraceId -> ...` not `(Maybe Word64`. So you either have to give it a name that says it's Datadog specific, or you need the caller to call a different (Datadogly-named) function to do the `TraceId -> Word64` conversion themselves. None of which I enjoyed very much.